### PR TITLE
Add preliminary support for datetime

### PR DESCRIPTION
**Pull request by saulshanabrook**
This is needed for https://github.com/Quansight/jupyterlab-omnisci/pull/85.

Closes https://github.com/Quansight/ibis-vega-transform/issues/7

### CODE REVIEW

File: ibis_vega_transform/vegaexpr.py [hunk 1]
```diff
@@ -11,11 +11,61 @@
...
 import ibis
 import ibis.expr.types as it
-import pytz
 from mypy_extensions import TypedDict
 from typing_extensions import Literal
 
 from altair_transform.utils import evaljs
+import altair_transform.utils._evaljs
+
+# Monkey patch altair_transform so that boolean operators  work on ibis expression
```
Comment (ian-r-rose) [inline ibis_vega_transform/vegaexpr.py]: That's kind of unfortunate that we have to do this. It looks like there are about 20 operators in the module -- do you think we'll have to cover all of them?

Comment (ian-r-rose) [reply ibis_vega_transform/vegaexpr.py]: Or might it be worthwhile to try to go a bit deeper into the `altair_transform` module and pull out the parser without the operators?

File: ibis_vega_transform/vegaexpr.py [hunk 2]
```diff
@@ -11,11 +11,61 @@
...
+
+
+def not_operator(value):
+    if isinstance(value, ibis.expr.types.ValueExpr):
```
Comment (ian-r-rose) [inline ibis_vega_transform/vegaexpr.py]: Are there any kinds of expressions we want to cover that are more basal than `ValueExpr`? I'm trying to understand the difference between that and `ibis.Expr`.

File: ibis_vega_transform/vegaexpr.py [hunk 3]
```diff
@@ -11,11 +11,61 @@
...
+        if isinstance(r, ibis.expr.types.ValueExpr):
+            return l & r
+        return r and l
+    return l and r
+
+
+def or_operator(l, r):
+    if isinstance(l, ibis.expr.types.ValueExpr):
+        if isinstance(r, ibis.expr.types.ValueExpr):
+            return l & r
```
Comment (ian-r-rose) [inline ibis_vega_transform/vegaexpr.py]: Is this a typo? Seems like it should be `l | r`

File: ibis_vega_transform/vegaexpr.py [hunk 4]
```diff
@@ -112,126 +162,161 @@ def now() -> float:
...
-        int(month) + 1,
-        int(day),
-        int(hour),
-        int(min),
-        int(sec),
-        int(millisec * 1000),
-        tzinfo=pytz.UTC,
-    )
+    raise NotImplementedError("utc()")
+    # return dt.datetime(
```
Comment (ian-r-rose) [inline ibis_vega_transform/vegaexpr.py]: dead code?

File: ibis_vega_transform/vegaexpr.py [hunk 5]
```diff
@@ -112,126 +162,161 @@ def now() -> float:
...
+        datetime(year(col), 0, 0, 0, 0, 0)
+        or
+        datetime(year(col), month(col), 0, 0, 0, 0)
+        etc.
+    """
+    if not hasattr(y, "_arg"):
+        raise NotImplementedError()
+    if not isinstance(y._arg, ibis.expr.operations.ExtractYear):
+        raise NotImplementedError()
+    date_column = y._arg.arg
```
Comment (ian-r-rose) [inline ibis_vega_transform/vegaexpr.py]: Yikes, where does `_arg` come from?

File: ibis_vega_transform/vegaexpr.py [hunk 6]
```diff
@@ -112,126 +162,161 @@ def now() -> float:
...
+    """
+    return not isinstance(v, ibis.expr.types.Expr) and v == 0
 
 
-def date(datetime: dt.datetime) -> int:
+def datetime(y, m, d, h, mi, s, ms):
+    """
+    This should use https://github.com/ibis-project/ibis/issues/386 if it gets implemented
+
+    Currently we only support use cases that can be turned into a truncate, which is:
```
Comment (ian-r-rose) [inline ibis_vega_transform/vegaexpr.py]: How serious do you think the `truncate` limitation is? Some example charts I've been playing with need non-truncation versions of this function.

Comment (ian-r-rose) [reply ibis_vega_transform/vegaexpr.py]: I'm wondering if we can support arbitrary timestamps here by using ibis timestamp literals, and then adding them together.

File: ibis_vega_transform/vegaexpr.py (hunk 2 above)
Comment (saulshanabrook) [reply ibis_vega_transform/vegaexpr.py]: I had a reason for switching it, to differentiate between ValueExpr and TableExpr, but I just tried reverting it back to the normal Expr and everything seems to work. So let's go with that for now. Thanks for picking up on this!

File: ibis_vega_transform/vegaexpr.py (hunk 3 above)
Comment (saulshanabrook) [reply ibis_vega_transform/vegaexpr.py]: Yep, thank you!

File: ibis_vega_transform/vegaexpr.py (hunk 4 above)
Comment (saulshanabrook) [reply ibis_vega_transform/vegaexpr.py]: Yep

File: ibis_vega_transform/vegaexpr.py (hunk 5 above)
Comment (saulshanabrook) [reply ibis_vega_transform/vegaexpr.py]: Who knows? I just found it on here. @xmnlab do you know of a better way to implement this sort of logic? I am trying to take apart an ibis expression into it's pieces to see how it was created.

File: ibis_vega_transform/vegaexpr.py (hunk 6 above)
Comment (saulshanabrook) [reply ibis_vega_transform/vegaexpr.py]: >  Some example charts I've been playing with need non-truncation versions of this function.

Could you point to an example, so I could take a look at it? I think we could get creative depending on the particular use case to try to work around other cases of it.

> I'm wondering if we can support arbitrary timestamps here by using ibis timestamp literals, and then adding them together.

Not sure I follow this, what would the resulting SQL look like? If omnisci had a `datetime` operator, then we wouldn't need any hacks here. cc @alexbaden who I think was looking into this

File: ibis_vega_transform/vegaexpr.py (hunk 1 above)
Comment (saulshanabrook) [reply ibis_vega_transform/vegaexpr.py]: Good point. I will take a look at the other operators to see if they work as is, or need to be patched as well. 

> Or might it be worthwhile to try to go a bit deeper into the altair_transform module and pull out the parser without the operators?

Yeah, that sounds good, long term. You OK with punting that for future work? Or would you like me to look into it for this PR?

Comment (saulshanabrook) [reply ibis_vega_transform/vegaexpr.py]: It looks like the other comparison operators (`<`, ...) work as is,  I am not sure if/how the [bitwise operators](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_Operators) are used in vega  expressions, so I am not sure how they should be translated.  I am inclined to add more overrides if we come across use cases which require them in the future.

**Pull request merged**